README.md
# OAuth2-Client

[![Build Status](https://travis-ci.org/aphillipo/oauth2-client.svg?branch=master)](https://travis-ci.org/scrogson/oauth2)
[![Coverage Status](https://coveralls.io/repos/aphillipo/oauth2-client/badge.svg?branch=master&service=github)](https://coveralls.io/github/scrogson/oauth2?branch=master)

> An Elixir OAuth2 Client 

## Install

```elixir
# mix.exs

def application do
  # Add the application to your list of applications.
  # This will ensure that it will be included in a release.
  [applications: [:logger, :oauth2client]]
end

defp deps do
  # Add the dependency
  [{:oauth2client, "~> 0.5"}]
end
```

## Usage

Current implemented strategies:

- Authorization Code
- Password
- Client Credentials

### Authorization Code Flow (AuthCode Strategy)

```elixir
# Initialize a client with client_id, client_secret, site, and redirect_uri.
# The strategy option is optional as it defaults to `OAuth2Client.Strategy.AuthCode`.

client = OAuth2Client.Client.new([
  strategy: OAuth2Client.Strategy.AuthCode, #default
  client_id: "client_id",
  client_secret: "abc123",
  site: "https://auth.example.com",
  redirect_uri: "https://example.com/auth/callback"
])

# Generate the authorization URL and redirect the user to the provider.
OAuth2Client.Client.authorize_url!(client)
# => "https://auth.example.com/oauth/authorize?client_id=client_id&redirect_uri=https%3A%2F%2Fexample.com%2Fauth%2Fcallback&response_type=code"

# Use the authorization code returned from the provider to obtain an access token.
token = OAuth2Client.Client.get_token!(client, code: "someauthcode")

# You can also use `OAuth2Client.Client.put_param/3` to update the client's `params`
# field. Example:
# token =
#   client
#   |> OAuth2Client.Client.put_param(:code, "someauthcode")
#   |> OAuth2Client.Client.get_token!()

# Use the access token to make a request for resources
resource = OAuth2Client.AccessToken.get!(token, "/api/resource").body
```

## Write Your Own Strategy

Here's an example strategy for GitHub:

```elixir
defmodule GitHub do
  use OAuth2Client.Strategy

  # Public API

  def client do
    OAuth2Client.Client.new([
      strategy: __MODULE__,
      client_id: "abc123",
      client_secret: "abcdefg",
      redirect_uri: "http://myapp.com/auth/callback",
      site: "https://api.github.com",
      authorize_url: "https://github.com/login/oauth/authorize",
      token_url: "https://github.com/login/oauth/access_token"
    ])
  end

  def authorize_url!(params \\ []) do
    client()
    |> put_param(:scope, "user,public_repo")
    |> OAuth2Client.Client.authorize_url!(params)
  end

  # you can pass options to the underlying http library via `options` parameter
  def get_token!(params \\ [], headers \\ [], options \\ []) do
    OAuth2Client.Client.get_token!(client(), params, headers, options)
  end

  # Strategy Callbacks

  def authorize_url(client, params) do
    OAuth2Client.Strategy.AuthCode.authorize_url(client, params)
  end

  def get_token(client, params, headers) do
    client
    |> put_header("Accept", "application/json")
    |> OAuth2Client.Strategy.AuthCode.get_token(params, headers)
  end
end
```

Here's how you'd use the example GitHub strategy:

Generate the authorize URL and redirect the client for authorization.

```elixir
GitHub.authorize_url!
```

Capture the `code` in your callback route on your server and use it to obtain an access token.

```elixir
token = GitHub.get_token!(code: code)
```

Use the access token to access desired resources.

```elixir
user = OAuth2Client.AccessToken.get!(token, "/user").body

# Or
case OAuth2Client.AccessToken.get(token, "/user") do
  {:ok, %OAuth2Client.Response{status_code: 401, body: body}} ->
    Logger.error("Unauthorized token")
  {:ok, %OAuth2Client.Response{status_code: status_code, body: user}} when status_code in [200..399] ->
    user
  {:error, %OAuth2Client.Error{reason: reason}} ->
    Logger.error("Error: #{inspect reason}")
end
```

## Examples

- [Authenticate with Github (OAuth2/Phoenix)](https://github.com/scrogson/oauth2_example)

## License

The MIT License (MIT)

Copyright (c) 2015 Sonny Scroggin

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Changelog

## v0.5.0 (2015-11-03)

### Improvements

* You can now request a refresh token with `OAuth2Client.AccessToken.refresh`. The `!` alternative is also available.
* Added `Bypass` for improved testability.
* `Plug` is no longer a direct dependency. It is only included as a test dependency through the `Bypass` library.
* `OAuth2Client.AccessToken` now supports `DELETE` requests with `delete` and `delete!`
* More tests!

### Bug Fixes

* Params are no longer sent in both the body and as a query string for `POST` requests with `OAuth2Client.Client.get_token`
* Responses will no longer be parsed automatically if the `content-type` is not supported by this lib. Registering custom parsers is a future goal for this library.
* Errors are now properly raised when they occur.

### Backwards Incompatible Changes

* `OAuth2Client.new/1` has been removed. Use `OAuth2Client.Client.new/1` instead.

Diff: https://github.com/scrogson/oauth2/compare/v0.4.0...v0.5.0

## v0.4.0 (2015-10-27)

### Additions/Improvements

* `OAuth2Client.AccessToken` now supports: `post`, `post!`, `put`, `put!`, `patch`, and `patch!`.
* Better documentation
* Test coverage improved

### Bug fixes

* Empty response bodies are no longer decoded

### Breaking changes

* `OAuth2Client.AccessToken.get!/4` now returns `OAuth2Client.Response{}` instead of just the parsed body.

### Aknowledgements

Thanks to @meatherly, @dejanstrbac, and @optikfluffel for their contributions!

Diff: https://github.com/scrogson/oauth2/compare/v0.3.0...v0.4.0

## v0.3.0 (2015-08-19)

Bump `Plug` dependency to `1.0`.

Diff: https://github.com/scrogson/oauth2/compare/v0.2.0...v0.3.0


## v0.2.0 (2015-07-13)

- `:erlang.now` was replaced with `:os.timestamp` for compatibility with Erlang 18
- You can now pass options to the `HTTPoison` library with `OAuth2Client.Client.get_token/4` and `OAuth2Client.Client.get_token!/4`

Diff: https://github.com/scrogson/oauth2/compare/v0.1.1...v0.2.0


## v0.1.1 (2015-04-18)

- Remove compilation warnings.
- Fix `request_body` function for `ClientCredentials`

Diff: https://github.com/scrogson/oauth2/compare/v0.1.0...v0.1.1


## v0.1.0 (2015-04-14)

This release bring breaking changes and more documentation.

Please see the [README](https://github.com/scrogson/oauth2/blob/v0.1.0/README.md) or [Hex Docs](http://hexdocs.pm/oauth2/0.1.0) for more details.

Diff: https://github.com/scrogson/oauth2/compare/v0.0.5...v0.1.0


## v0.0.5 (2015-04-11)

- Handles Facebooks `expires` key for Access Tokens.
- Ensure the token type defaults to 'Bearer' when it is not present.

Diff: https://github.com/scrogson/oauth2/compare/0.0.3...v0.0.5


## v0.0.3 (2015-01-12)

- Relax version requirements for Poison.

## v0.0.2 (2015-01-10)

This release brings Password and Client Credentials strategies.

http://tools.ietf.org/html/draft-ietf-oauth-v2-15#section-4.3
http://tools.ietf.org/html/draft-ietf-oauth-v2-15#section-4.4

## v0.0.1 (2014-12-07)

Initial release.

This initial release includes a functional authorization code strategy: http://tools.ietf.org/html/draft-ietf-oauth-v2-15#section-4.1


lib/oauth2client/access_token.ex
defmodule OAuth2Client.AccessToken do
  @moduledoc """
  This module defines the `OAuth2Client.AccessToken` struct and provides functionality
  to make authorized requests to an OAuth2Client provider using the AccessToken
  returned by the provider.

  The `OAuth2Client.AccessToken` struct is created for you when you use the
  `OAuth2Client.Client.get_token`

  ### Notes

  * If a full url is given (e.g. "http://www.example.com/api/resource") then it
  will use that otherwise you can specify an endpoint (e.g. "/api/resource") and
  it will append it to the `Client.site`.

  * The headers from the `Client.headers` are appended to the request headers.

  ### Examples

  ```
  token =  OAuth2Client.AccessToken.new("abc123", %OAuth2Client.Client{site: "www.example.com"})

  case OAuth2Client.AccessToken.get(token, "/some/resource") do
    {:ok, %OAuth2Client.Response{status_code: 401}} ->
      "Not Good"
    {:ok, %OAuth2Client.Response{status_code: status_code, body: body}} when status_code in [200..299] ->
      "Yay!!"
    {:error, %OAuth2Client.Error{reason: reason}} ->
      reason
  end

  response = OAuth2Client.AccessToken.get!(token, "/some/resource")

  response = OAuth2Client.AccessToken.post!(token, "/some/other/resources", %{foo: "bar"})
```

  """

  import OAuth2Client.Util

  alias OAuth2Client.Error
  alias OAuth2Client.Client
  alias OAuth2Client.Request
  alias OAuth2Client.Response
  alias OAuth2Client.AccessToken

  @standard ["access_token", "refresh_token", "expires_in", "token_type"]

  @type access_token  :: binary
  @type refresh_token :: binary
  @type expires_at    :: integer
  @type token_type    :: binary
  @type other_params  :: %{}
  @type body          :: binary | %{}

  @type t :: %__MODULE__{
              access_token:  access_token,
              refresh_token: refresh_token,
              expires_at:    expires_at,
              token_type:    token_type,
              other_params:  other_params,
              client:        Client.t}

  defstruct access_token: "",
            refresh_token: nil,
            expires_at: nil,
            token_type: "Bearer",
            other_params: %{},
            client: nil

  @doc """
  Returns a new `OAuth2Client.AccessToken` struct given the access token `string`.

  ### Example

  ```
  iex(1)> OAuth2Client.AccessToken.new("abc123", %OAuth2Client.Client{})
  %OAuth2Client.AccessToken{access_token: "abc123",
   client: %OAuth2Client.Client{authorize_url: "/oauth/authorize", client_id: "",
    client_secret: "", headers: [], params: %{}, redirect_uri: "", site: "",
    strategy: OAuth2Client.Strategy.AuthCode, token_method: :post,
    token_url: "/oauth/token"}, expires_at: nil, other_params: %{},
   refresh_token: nil, token_type: "Bearer"}
  ```

  """
  @spec new(binary, Client.t) :: t
  def new(token, client) when is_binary(token) do
    new(%{"access_token" => token}, client)
  end

  @doc """
  Same as `new/2` except that the first arg is a `map`.

  Note if giving a map, please be sure to make the key a `string` no an `atom`.

  This is used by `OAuth2Client.Client.get_token/4` to create the `OAuth2Client.AccessToken` struct.

  ### Example

  ```
  iex(1)> OAuth2Client.AccessToken.new(%{"access_token" => "abc123"}, %OAuth2Client.Client{})
   %OAuth2Client.AccessToken{access_token: "abc123",
    client: %OAuth2Client.Client{authorize_url: "/oauth/authorize", client_id: "",
     client_secret: "", headers: [], params: %{}, redirect_uri: "", site: "",
     strategy: OAuth2Client.Strategy.AuthCode, token_method: :post,
     token_url: "/oauth/token"}, expires_at: nil, other_params: %{},
    refresh_token: nil, token_type: "Bearer"}
  ```
  """
  def new(response, client) do
    {std, other} = Dict.split(response, @standard)

    struct __MODULE__, [
      access_token:  std["access_token"],
      refresh_token: std["refresh_token"],
      expires_at:    (std["expires_in"] || other["expires"]) |> expires_at(),
      token_type:    std["token_type"] |> normalize_token_type(),
      other_params:  other,
      client:        client]
  end

  @doc """
  Makes a `GET` request to the given `url` using the `OAuth2Client.AccessToken`
  struct.
  """
  @spec get(t, binary, Client.headers, Keyword.t) :: {:ok, Response.t} | {:error, Error.t}
  def get(token, url, headers \\ [], opts \\ []),
    do: request(:get, token, url, "", headers, opts)

  @doc """
  Same as `get/4` but returns a `OAuth2Client.Response` or `OAuth2Client.Error` exception if
  the request results in an error.
  """
  @spec get!(t, binary, Client.headers, Keyword.t) :: Response.t | Error.t
  def get!(token, url, headers \\ [], opts \\ []),
    do: request!(:get, token, url, "", headers, opts)

  @doc """
  Makes a `PUT` request to the given `url` using the `OAuth2Client.AccessToken`
  struct.
  """
  @spec put(t, binary, body, Client.headers, Keyword.t) :: {:ok, Response.t} | {:error, Error.t}
  def put(token, url, body \\ "", headers \\ [], opts \\ []),
    do: request(:put, token, url, body, headers, opts)

  @doc """
  Same as `put/5` but returns a `OAuth2Client.Response` or `OAuth2Client.Error` exception if
  the request results in an error.

  An `OAuth2Client.Error` exception is raised if the request results in an
  error tuple (`{:error, reason}`).
  """
  @spec put!(t, binary, body, Client.headers, Keyword.t) :: Response.t | Error.t
  def put!(token, url, body \\ "", headers \\ [], opts \\ []),
    do: request!(:put, token, url, body, headers, opts)

  @doc """
  Makes a `PATCH` request to the given `url` using the `OAuth2Client.AccessToken`
  struct.
  """
  @spec patch(t, binary, body, Client.headers, Keyword.t) :: {:ok, Response.t} | {:error, Error.t}
  def patch(token, url, body \\ "", headers \\ [], opts \\ []),
    do: request(:patch, token, url, body, headers, opts)

  @doc """
  Same as `patch/5` but returns a `OAuth2Client.Response` or `OAuth2Client.Error` exception if
  the request results in an error.

  An `OAuth2Client.Error` exception is raised if the request results in an
  error tuple (`{:error, reason}`).
  """
  @spec patch!(t, binary, body, Client.headers, Keyword.t) :: Response.t | Error.t
  def patch!(token, url, body \\ "", headers \\ [], opts \\ []),
    do: request!(:patch, token, url, body, headers, opts)

  @doc """
  Makes a `POST` request to the given URL using the `OAuth2Client.AccessToken`.
  """
  @spec post(t, binary, body, Client.headers, Keyword.t) :: {:ok, Response.t} | {:error, Error.t}
  def post(token, url, body \\ "", headers \\ [], opts \\ []),
    do: request(:post, token, url, body, headers, opts)

  @doc """
  Same as `post/5` but returns a `OAuth2Client.Response` or `OAuth2Client.Error` exception
  if the request results in an error.

  An `OAuth2Client.Error` exception is raised if the request results in an
  error tuple (`{:error, reason}`).
  """
  @spec post!(t, binary, body, Client.headers, Keyword.t) :: Response.t | Error.t
  def post!(token, url, body \\ "", headers \\ [], opts \\ []),
    do: request!(:post, token, url, body, headers, opts)

  @doc """
  Makes a `DELETE` request to the given URL using the `OAuth2Client.AccessToken`.
  """
  @spec delete(t, binary, body, Client.headers, Keyword.t) :: {:ok, Response.t} | {:error, Error.t}
  def delete(token, url, body \\ "", headers \\ [], opts \\ []),
    do: request(:delete, token, url, body, headers, opts)

  @doc """
  Same as `delete/5` but returns a `OAuth2Client.Response` or `OAuth2Client.Error` exception
  if the request results in an error.

  An `OAuth2Client.Error` exception is raised if the request results in an
  error tuple (`{:error, reason}`).
  """
  @spec delete!(t, binary, body, Client.headers, Keyword.t) :: Response.t | Error.t
  def delete!(token, url, body \\ "", headers \\ [], opts \\ []),
    do: request!(:delete, token, url, body, headers, opts)

  @doc """
  Makes a request of given type to the given URL using the `OAuth2Client.AccessToken`.
  """
  @spec request(atom, t, binary, body, Client.headers, Keyword.t) :: {:ok, Response.t} | {:error, Error.t}
  def request(method, token, url, body \\ "", headers \\ [], opts \\ []) do
    url = process_url(token, url)
    headers = req_headers(token, headers)

    case Request.request(method, url, body, headers, opts) do
      {:ok, response} -> {:ok, response}
      {:error, error} -> {:error, error}
    end
  end

  @doc """
  Same as `request/6` but returns `OAuth2Client.Response` or raises an error if an
  error occurs during the request.

  An `OAuth2Client.Error` exception is raised if the request results in an
  error tuple (`{:error, reason}`).
  """
  @spec request!(atom, t, binary, body, Client.headers, Keyword.t) :: Response.t | Error.t
  def request!(method, token, url, body \\ "", headers \\ [], opts \\ []) do
    case request(method, token, url, body, headers, opts) do
      {:ok, response} -> response
      {:error, error} -> raise error
    end
  end

  @doc """
  Gets a new `AccessToken` by making a request using the refresh_token

  Returns an `AccessToken` struct that can then be used to access the resource API.
  """
  @spec refresh(t, Client.params, Client.headers, Keyword.t) :: {:ok, AccessToken.t} | {:error, Error.t}
  def refresh(token, params \\ [], headers \\ [], opts \\ [])
  def refresh(%{refresh_token: nil}, _params, _headers, _opts) do
    {:error, %Error{reason: "Refresh token not available."}}
  end
  def refresh(%{refresh_token: refresh_token, client: client}, params, headers, opts) do
    refresh =
      %{client | strategy: OAuth2Client.Strategy.Refresh}
      |> Client.put_param(:refresh_token, refresh_token)

    case Client.get_token(refresh, params, headers, opts) do
      {:ok, token}    -> {:ok, %{token | client: client}}
      {:error, error} -> {:error, error}
    end
  end

  @doc """
  Calls `refresh/3` but raises `Error` if there an error occurs.
  """
  @spec refresh!(t, Client.params, Client.headers, Keyword.t) :: AccessToken.t | Error.t
  def refresh!(token, params \\ [], headers \\ [], opts \\ []) do
    case refresh(token, params, headers, opts) do
      {:ok, token} -> token
      {:error, error} -> raise error
    end
  end


  @doc """
  Determines if the access token will expire or not.

  Returns `true` unless `expires_at` is `nil`.
  """
  @spec expires?(OAuth2Client.AccessToken.t) :: boolean
  def expires?(%AccessToken{expires_at: nil} = _token), do: false
  def expires?(_), do: true

  @doc """
  Determines if the access token has expired.
  """
  def expired?(token) do
    expires?(token) && unix_now > token.expires_at
  end

  @doc """
  Returns a unix timestamp based on now + expires_at (in seconds).
  """
  def expires_at(nil), do: nil
  def expires_at(val) when is_binary(val) do
    {int, _} = Integer.parse(val)
    int
  end
  def expires_at(int), do: unix_now + int

  defp process_url(token, url) do
    case String.downcase(url) do
      <<"http://"::utf8, _::binary>> -> url
      <<"https://"::utf8, _::binary>> -> url
      _ -> token.client.site <> url
    end
  end

  defp normalize_token_type(nil), do: "Bearer"
  defp normalize_token_type("bearer"), do: "Bearer"
  defp normalize_token_type(string), do: string

  defp req_headers(token, headers) do
    [{"Authorization", "#{token.token_type} #{token.access_token}"} | headers] ++ token.client.headers
  end
end
